resources: GPUFS Ubuntu22.04 ROCm 5.4.2 disk image

abmerop opened the pull request:
WIP: This is missing the kernel and vBIOS resources. Discuss in github.

Contains everything needed to create a GPUFS disk image from scratch. Currently we support ROCm 4.2 in the util directory, however everything up to the currenct ROCm should work. This provides a way to be able to create a disk image which simulates a more recent GPU software stack.

With this disk image we can run most ROCm 5.4 and some PyTorch, and Tensorflow applications.

abmerop committed:
resources: GPUFS Ubuntu22.04 ROCm 5.4.2 disk image

Contains everything needed to create a GPUFS disk image from scratch.
Currently we support ROCm 4.2 in the util directory, however everything
up to the currenct ROCm should work. This provides a way to be able to
create a disk image which simulates a more recent GPU software stack.

With this disk image we can run ROCm 5.4 as well as some PyTorch and
Tensorflow applications.

There are instructions to obtain the Linux kernel and it can be
downloaded as a resource.

abmerop commented:
somewhat of a copy of #6 but after two weeks made my own.

Open questions: How to handle the vBIOS( "ROM") and very specific kernel . Put some notes in the README.md.

We could have the vBIOS be a resource (* requires gem5 code changes) or simply place them in the packer files directory.


FYI this image runs square, hacc (to make sure MPI installed), pytorch, and tensorflow (simple lin regs), all of the hip examples, night, weekly tests.

abmerop commented:
Updated pytorch and tensorflow installs to use sudo.  Since gem5 runs as root, this should resolve any issue where they are not found.

abmerop commented:
Add to TODO.  Not sure if anyone knows what service takes 2 minutes to load. Not great for simulation outside of KVM:

A start job is running for Wait for… to be Configured (53s / no limit)

abmerop commented:
Added vbios for vega10 and mi200 to commit.  No plans to include mi100 for two reasons:

1) mi200 is newer..
2) mi100 has an extra register file that gem5 doesn't model and was immediately removed in mi200

powerjg commented:
A couple of quick comments:
1. The thing that take 2 minutes is likely the network... we need to figure out how to disable that. I think we had something working in the past, but we need to look into it again.
2. Thanks for putting this together. We (UCD) are going to take a concerted effort to update all of the disk images in the next few weeks. Unless this is blocking you, we'll review this change then.

abmerop commented:
Not blocking me.  There's been some interest in running PyTorch/TF so I just wanted to get something compatible out there.  For now those folks should be able to pull this from my github

ivanaamit commented:
We (UCD) will be working on Disk images for the next couple of weeks, and we'll get back to you when we have updates.

abmerop committed:
resources: Remove old ROCm 4.2 disk image packer files

Remove old Ubuntu 18.04 and ROCm 4.2 disk image files along with the
README. The contents of the README should be moved to the website.

abmerop commented:
This is ready to go now, IMO.

Fixed the 2 minute boot delay with this:

`sudo systemctl mask open-iscsi.service iscsid.service iscsid.socket`

after trying every single other option here: https://askubuntu.com/questions/972215/a-start-job-is-running-for-wait-for-network-to-be-configured-ubuntu-server-17-1

Maybe one nice-to-have feature would be to extract the kernel. Let me know.

mattsinc commented on src/x86-ubuntu-22.04-rocm-5.4.2/README.md:
```diff
@@ -0,0 +1,97 @@
+   - build/VEGA_X86/gem5.opt configs/example/gpufs/vega10_kvm.py --disk-image gem5-resources/src/x86-ubuntu-22.04-rocm-5.4.2/qemu_files/ubuntu2204-rocm542.img  --kernel gem5-resources/src/x86-ubuntu-22.04-rocm-5.4.2/qemu_files/vmlinux-5.15.0-75-generic --gpu-mmio-trace gem5-resources/src/gpu-fs/vega_mmio.log --app some/where/tensorflow_app.sh
+
+
+## Coming soon (TM): stdlib GPUFS config
```
I will leave it up to the others if we want this here, or it should just be removed until it exists.

mattsinc replied:
My comment is the same here

mattsinc reviewed (dismissed):
This seems good to me, amazing job!  The others should probably ok too, before we merge in though.

mattsinc reviewed (dismissed):
Ok on my end, but Bobby or Jason should approve too.

abmerop commented:
FYI: Planning to make some edits to the README file, will re-request a review once that is done

BobbyRBruce reviewed (dismissed):
I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:

1. Is this modifying any existing disk images we have in our bucket?
2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.

abmerop commented:
> I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> 
> 1. Is this modifying any existing disk images we have in our bucket?
> 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.

Just seeing this, sorry for the delay.


1) This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space. 
2) Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
3) My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2


....   I don't know why the diff is show 560 files. Closing #17

mattsinc commented:
> > I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> > 
> > 1. Is this modifying any existing disk images we have in our bucket?
> > 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> > 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.
> 
> Just seeing this, sorry for the delay.
> 
>     1. This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space.
> 
>     2. Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
> 
>     3. My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2
> 
> 
> .... I don't know why the diff is show 560 files. Closing #17

As far as I can tell this hasn't changed, right?  If so, I think we are waiting on Bobby to approve?

abmerop commented:
> > > I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> > > 
> > > 1. Is this modifying any existing disk images we have in our bucket?
> > > 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> > > 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.
> > 
> > 
> > Just seeing this, sorry for the delay.
> > ```
> > 1. This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space.
> > 
> > 2. Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
> > 
> > 3. My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > .... I don't know why the diff is show 560 files. Closing #17
> 
> As far as I can tell this hasn't changed, right? If so, I think we are waiting on Bobby to approve?

I just made very minor edits to the README and now rebased (correctly).  Nothing else has changed

mattsinc commented:
> > > > I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> > > > 
> > > > 1. Is this modifying any existing disk images we have in our bucket?
> > > > 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> > > > 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.
> > > 
> > > 
> > > Just seeing this, sorry for the delay.
> > > ```
> > > 1. This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space.
> > > 
> > > 2. Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
> > > 
> > > 3. My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > .... I don't know why the diff is show 560 files. Closing #17
> > 
> > 
> > As far as I can tell this hasn't changed, right? If so, I think we are waiting on Bobby to approve?
> 
> I just made very minor edits to the README and now rebased (correctly). Nothing else has changed

I'll let Bobby approve then since these questions were his.

BobbyRBruce commented:
> > > > > I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> > > > > 
> > > > > 1. Is this modifying any existing disk images we have in our bucket?
> > > > > 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> > > > > 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.
> > > > 
> > > > 
> > > > Just seeing this, sorry for the delay.
> > > > ```
> > > > 1. This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space.
> > > > 
> > > > 2. Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
> > > > 
> > > > 3. My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > .... I don't know why the diff is show 560 files. Closing #17
> > > 
> > > 
> > > As far as I can tell this hasn't changed, right? If so, I think we are waiting on Bobby to approve?
> > 
> > 
> > I just made very minor edits to the README and now rebased (correctly). Nothing else has changed
> 
> I'll let Bobby approve then since these questions were his.

It seems like this is really targetting the v23.1 release. In which case i'll hold off from merging this until the release.

BobbyRBruce commented:
> > I've got no complaints here though I have forgotten what the state of things is. Some questions purely for my benefit:
> > 
> > 1. Is this modifying any existing disk images we have in our bucket?
> > 2. Do we need to add a new disk image? I assume you want to pull this to use in simulations? If so, just let me and @Harshil2107 know. Sending us a pre-build disk image would be beneficial.
> > 3. You're targeting the `stable` branch. So, this change is ok for v23.0? If intended for v23.1, no big deal. You can wait until that release has occured _or_ retarget the `develop` branch instead.
> 
> Just seeing this, sorry for the delay.
> 
> 1. This is a completely standalone image. I was originally trying to reuse something like gabps but i did not get a response and ROCm + PyTorch require at least 30 GB more disk space.
> 2. Yes, because of above.  Let me know how I can send a pre-built disk.  You will also want the ubuntu kernel.
> 3. My intention is to wait for v23.1.   There are many fixes since v23.0 for ROCm 5.4.2
> 
> .... I don't know why the diff is show 560 files. Closing #17

For (2): Yes, please send us a pre-built. How? I don't have a good answer to that. Do we have somewhere you can put this so we can download at UC Davis? A dropbox link? Anything works for us really.

abmerop commented:
Should we push this now that v23.1 is released?

I will have to delay posting the disk image.  It is difficult for me to upload it due to the size and slow upload speed.

mattsinc commented on src/gpu-fs/README.md:
```diff
@@ -1,133 +0,0 @@
----
```
Is the whole gpu-fs folder deleted now in favor of the x86-ubuntu... folder?  If so that's fine.  If not it's a bit strange for the README to disappear here.

abmerop replied:
Yes, that's correct.  There is a new README in the new folder.


I'd like to push this immediately as there are already users complaining that the ROCm 4.2 disk image does not work with the ROCm 5.4 docker image to build binaries

mattsinc replied:
This is fine with me ... I guess we should give @BobbyRBruce a couple hours to reply though

powerjg replied:
Bobby is out sick today, unfortunately.

mattsinc reviewed (approved):
As far as I can tell, this is the same as before with maybe minor README modifications, so I'm approving.  However @BobbyRBruce had pending questions before, so I guess Bobby should let us know if we should not move in.

powerjg reviewed (approved):
I think it's fine to merge this now. Getting the disk image uploaded is also important, but that is going to take finding a way for you to upload it.

BobbyRBruce commented:
Hey @abmerop : Can you upload the resources updated via this source change to this box.com folder? : https://ucdavis.app.box.com/f/c3074cc527f842d5a7391f86e4b9e63a.

If you do so we can upload them to our cloud bucket and fully incorporate them into gem5 resources.

abmerop commented:
> Hey @abmerop : Can you upload the resources updated via this source change to this box.com folder? : https://ucdavis.app.box.com/f/c3074cc527f842d5a7391f86e4b9e63a.
> 
> If you do so we can upload them to our cloud bucket and fully incorporate them into gem5 resources.

Hi Bobby.  It says the image is too big.  I will attempt to compress it but there is essentially no free space left on the image that would be easily compressed. ROCm install is fairly large (I think 10GB?) with all of the 1st party libraries and such.  PyTorch and Tensorflow package *another* ROCm install with their installs to avoid dependency issues which means it's essentially installed 3 times.  The total size therefore is around 35GB. It is possible to create an image without ML libraries to get a smaller image. Those libraries could be part of a separate optional image too if that is useful.

BobbyRBruce commented:
> > Hey @abmerop : Can you upload the resources updated via this source change to this box.com folder? : https://ucdavis.app.box.com/f/c3074cc527f842d5a7391f86e4b9e63a.
> > If you do so we can upload them to our cloud bucket and fully incorporate them into gem5 resources.
> 
> Hi Bobby. It says the image is too big. I will attempt to compress it but there is essentially no free space left on the image that would be easily compressed. ROCm install is fairly large (I think 10GB?) with all of the 1st party libraries and such. PyTorch and Tensorflow package _another_ ROCm install with their installs to avoid dependency issues which means it's essentially installed 3 times. The total size therefore is around 35GB. It is possible to create an image without ML libraries to get a smaller image. Those libraries could be part of a separate optional image too if that is useful.

Was there any success in compressing the image?

abmerop commented:
> > > Hey @abmerop : Can you upload the resources updated via this source change to this box.com folder? : https://ucdavis.app.box.com/f/c3074cc527f842d5a7391f86e4b9e63a.
> > > If you do so we can upload them to our cloud bucket and fully incorporate them into gem5 resources.
> > 
> > 
> > Hi Bobby. It says the image is too big. I will attempt to compress it but there is essentially no free space left on the image that would be easily compressed. ROCm install is fairly large (I think 10GB?) with all of the 1st party libraries and such. PyTorch and Tensorflow package _another_ ROCm install with their installs to avoid dependency issues which means it's essentially installed 3 times. The total size therefore is around 35GB. It is possible to create an image without ML libraries to get a smaller image. Those libraries could be part of a separate optional image too if that is useful.
> 
> Was there any success in compressing the image?

Yes, the compressed image is 11GB.  I've uploaded that compressed image and the corresponding Linux kernel to that Box link on Friday
